### PR TITLE
Updates

### PR DESCRIPTION
**Pull request by whitequark**
See the individual commits.

### CODE REVIEW

**Comment by whitequark** [comment]
Things this PR doesn't implement:
  * `Consumer`/`Producer` wrappers. I am personally not expecting to actually use those; and they can be implemented using only the public API. I think it's fine to defer them until someone actually wants those. There are also some subtle points wrt implementing the Iterator trait, which I'm not deeply interested in exploring right now.
  * Forcible unwinding. To begin with, we need to figure out how to configure the behavior on drop. The reason is that, if we forcibly unwind on drop with with -Cpanic=abort, this means that all memory leaks become mandatory aborts, which is undesirable. Printing a message to stderr is also not always desirable. Adding some sort of type parameter to configure this, or even a `new` parameter, clutters the interface. I think it is an interesting direction to explore, but it isn't yet fully clear how to do it.

**Comment by whitequark** [comment]
OK, this is ready for merging. OR1K support pending, once I figure out how to wrestle the OR1K rustc target into submission so that it works with cargo.

File: src/generator.rs [hunk 1]
```diff
@@ -24,18 +23,20 @@ pub enum State {
...
+/// a value each time.
 ///
-/// It implements the Iterator trait. The first time `next()` is called,
-/// the function is called as `f(yielder)`; every time `next()` is called afterwards,
-/// the function is resumed. In either case, it runs until it suspends its execution
-/// through `yielder.generate(val)`), in which case `next()` returns `Some(val)`, or
-/// returns, in which case `next()` returns `None`. `next()` will return `None`
-/// every time after that.
+/// The first time `resume(input0)` is called, the function is called as `f(yielder, input0)`.
+/// It runs until it suspents its execution through `yielder.generate(output0)`, after which
```
Comment (nathan7) [inline src/generator.rs]: s/suspents/suspends/

Comment (whitequark) [inline src/generator.rs]: fixed

File: src/generator.rs [hunk 2]
```diff
@@ -77,36 +79,70 @@ impl<Item, Stack> Generator<Item, Stack>
...
+        if val.is_some() { self.state = State::Runnable }
+
+        val
+      }
+      State::Unavailable => None
+    }
+  }
+
   /// Returns the state of the generator.
   pub fn state(&self) -> State { self.state }
```
Comment (nathan7) [inline src/generator.rs]: prime `#[inline]` candidate

File: src/generator.rs [hunk 3]
```diff
@@ -124,72 +160,41 @@ impl<Item, Stack> Generator<Item, Stack>
...
+      self.context = new_context;
+      // However, between this point and the next time we enter generate_bare
       // the generator cannot be moved, as a &mut Generator is necessary
       // to resume the generator function.
+      input
     }
   }
 
-  /// Suspends the generator and returns `Some(item)` from the `next()`
+  /// Suspends the generator and returns `Some(item)` from the `call()`
```
Comment (nathan7) [inline src/generator.rs]: s/call()/resume()/

File: src/arch/x86.rs [hunk 4]
```diff
@@ -28,6 +28,14 @@
 //   after. A .cfi_def_* pseudoinstruction changes the CFA value similarly.
 // * Simulating return is as easy as restoring register values from the CFI table
 //   and then setting stack pointer to CFA.
+//
+// A high-level overview of the function of the trampolines is:
+// * The 2nd init trampoline puts a controlled value (written in swap to `new_cfa`)
+//   into %ebx.
+// * The 1st init trampoline tells the unwinder to set %esp to %ebx, thus continuing
+//   unwinding at the swap call site instead of falling off the end of context stack.
+// * The 1st init trampoline together with the swap trampoline also restore %erp
```
Comment (nathan7) [inline src/arch/x86.rs]: that's `%esp`, I hope?

**Comment by nathan7** [comment]
Merged :heart:

**Pull request closed without merging**